### PR TITLE
Add EpicWithSpecialHeader

### PR DESCRIPTION
**Pull request by tjmw**
## What does this change?

This PR adds a new component `EpicWithHeaderImage`. The new component is implemented on top of the existing Epic component and adds a new header section including a headshot and some additional text.

This PR supersedes #222. It turned out the image marketing wanted to use was very text heavy, which didn't feel like a great fit (accessibility concerns, as well as general inflexibility of delivering text in an image).

Note: The entire new top section is hidden on mobile.

We’re treating our first usage of the epic with an image as an experiment, so short term we’ll hardcode the text and image to avoid the overhead of adding proper image support (via the picker). If the test goes well we can make all this a properly supported part of the Epic component.

https://trello.com/c/xODg9J4r/1217-new-epic-template-with-image

## How to test

In Storybook, there's a new story EpicWithHeaderImage under WorkInProgress/EndOfArticle.

## How can we measure success?

We'll be running a test to validate this new feature.

## Images

Screenshot 2021-09-14 at 15 15 34

## Accessibility

- [ ] [Tested with screen reader](https://github.com/guardian/source/blob/main/docs/06-accessibility.md#screen-readers)
- [ ] [Navigable with keyboard](https://github.com/guardian/source/blob/main/docs/06-accessibility.md#keyboard-navigation)
- [ ] [Colour contrast passed](https://github.com/guardian/source/blob/main/docs/06-accessibility.md#colour-contrast)
- [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/source/blob/main/docs/06-accessibility.md#use-of-colour)

### CODE REVIEW

File: src/EpicWithHeaderImage/index.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+};
+
+// We use a picture element here with an empty image for < tablet so that we
+// don't load an image unnecessarily when we're not going to render it (which would be
+// the case with an img tag)
+const HeaderSection: React.FC = () => (
+    <div css={headerStyles.container}>
+        <div css={headerStyles.leftContainer}>
+            <span css={headerStyles.text}>
+                You&#8217;re powering open,&nbsp;independent journalism
```
Comment (tjmw) [inline src/EpicWithHeaderImage/index.tsx:64]: I used a non-breaking space here to keep `open, independent` on the same line. Interested in thoughts on that. It seems to have the desired effect on the browsers/widths I've tested on.

Comment (adamnfish) [reply src/EpicWithHeaderImage/index.tsx:64]: Sounds good if we've checked it doesn't cause problems at any width. A more conservative approach in general is to add word breaks elsewhere. This mean the browser **favours** other splits rather than **forcing** this lack of split, which gives it the flexibility to make this break if doing so is completely necessary.

If this isn't being shown at mobile widths then I imagine this is fine!

File: src/BrazeEndOfArticleComponent.ts [hunk 2]
```diff
@@ -7,12 +7,13 @@ import {
...
-
 import { COMPONENT_NAME as AU_NEWSLETTER_EPIC_NAME, AUNewsletterEpic } from './AUNewsletterEpic';
-
 import { COMPONENT_NAME as UK_NEWSLETTER_EPIC_NAME, UKNewsletterEpic } from './UKNewsletterEpic';
-
 import { COMPONENT_NAME as EPIC_NAME, Epic } from './Epic';
+import {
+    COMPONENT_NAME as EPIC_WITH_HEADER_IMAGE_NAME,
+    EpicWithHeaderImage,
+} from './EpicWithHeaderImage';
```
Comment (adamnfish) [inline src/BrazeEndOfArticleComponent.ts:16]: Super minor point, but this name kind of doesn't apply anymore! Now it's an epic with a bespoke title header. It does have an image in so this isn't totally misleading! Fine for this test, maybe let's double check our thoughts if we go any further with templating this approach?

File: src/Epic/index.tsx [hunk 3]
```diff
@@ -23,7 +23,7 @@ const styles = {
         max-width: 620px;
     `,
     epicContainer: css`
-        padding: 4px 8px 12px;
+        padding: ${space[1]}px ${space[2]}px ${space[3]}px;
```
Comment (adamnfish) [inline src/Epic/index.tsx:26]: nice improvement

File: src/Epic/index.tsx [hunk 4]
```diff
@@ -90,6 +92,8 @@ export const Epic: React.FC<EpicProps> = (props: EpicProps) => {
         <ThemeProvider theme={brand}>
             <div css={styles.epicWrapper}>
                 <section css={styles.epicContainer}>
+                    {children}
```
Comment (adamnfish) [inline src/Epic/index.tsx:95]: Is there a better name than `children` here? Children is fine and technically accurate, but it doesn't convey the context around how this element is used. This has semantic significance here from where in the markup it is placed that we could convey with this name. Perhaps this is part of a wider interface that I haven't spotted, in which case the specific context here doesn't apply so we shouldn't pick a specific name? If not, perhaps mentioning that this appears as a header would be helpful?

File: src/EpicWithHeaderImage/index.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                componentName={args.componentName}
+                subscribeToNewsletter={() => Promise.resolve()}
+            />
+        </StorybookWrapper>
+    );
+};
+
+export const DefaultStory = StoryTemplate.bind({});
+
+DefaultStory.args = {
```
Comment (adamnfish) [inline src/EpicWithHeaderImage/index.stories.tsx:81]: Should this default story include header text? II guess it's missing because it's hard-coded in the template, in which case that makes sense.

File: src/storybookCommon/argTypes.ts [hunk 6]
```diff
@@ -29,3 +29,23 @@ export const ophanComponentIdArgType = {
             '`rc_upsell_test1_variant`.',
     },
 };
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (adamnfish) [inline src/storybookCommon/argTypes.ts:33]: Why is this required here? It looks like we should know the types of what we're returning?

**Review by adamnfish — APPROVED** [review]
Couple of little Qs, but looks great!

File: src/BrazeEndOfArticleComponent.ts (hunk 2 above)
Comment (tjmw) [reply src/BrazeEndOfArticleComponent.ts:16]: Ha yeah that's a good point, I think I'd managed to convince myself it was okay because like you say there is still an image in there. I'll give some thought to what a good name would be. I'm hoping this template will be short lived so sticking with the current name doesn't seem _that_ bad, would be nice to do better though.

File: src/Epic/index.tsx (hunk 4 above)
Comment (tjmw) [reply src/Epic/index.tsx:95]: Good questions. The `children` prop is a [React thing](https://reactjs.org/docs/composition-vs-inheritance.html#containment), so if you specify your jsx like this:

```jsx
<A>
  <B />
</A>
```

Then the inner content of `A` is made available to `A` as `children`. But there are other options we could use here to give it a better name. We could capture the arg into a different name using destructuring like this:

```jsx
const A = ({ children: specialHeader ) => { ... }
```

Or alternatively pass the special header like any other prop, like this:

```jsx
<A specialHeader={<B />} />
```

(That last example is how you'd have to do things if you wanted to support a custom header _and_ footer for example).

Any thoughts on which of those might be preferable?

File: src/storybookCommon/argTypes.ts (hunk 6 above)
Comment (tjmw) [reply src/storybookCommon/argTypes.ts:33]: Good spot, it's not needed (it was left over from a previous iteration of this work).

File: src/EpicWithHeaderImage/index.stories.tsx (hunk 5 above)
Comment (tjmw) [reply src/EpicWithHeaderImage/index.stories.tsx:81]: Yeah that's correct. The original/core epic template support supports a simple text header which is provided dynamically from Braze. Here we're omitting it as we're using the hardcoded new special header section.

File: src/EpicWithHeaderImage/index.tsx (hunk 1 above)
Comment (tjmw) [reply src/EpicWithHeaderImage/index.tsx:64]: Also a good point 👍 

> is to add word breaks elsewhere

You mean with `<wbr>`? I've not used those much before, I'll take a look.

I just tried removing the nbsp and the browser is wrapping things exactly the same, maybe this was making a difference in a previous version where this top section was laid out slightly differently. So I'm tempted to remove it. (In a future where that text is dynamically provided we probably wouldn't have that level of control, so maybe better to let the browser get on with it).

Comment (adamnfish) [reply src/EpicWithHeaderImage/index.tsx]: Yeah a `<wbr>` tag is one way, IIRC there's an entity for it as well*. It's well-supported in theory but I don't know how strongly browsers would stick to it these days. My experience with this is from a few years back!

* Like me, this page is very old but has some information on the options
    https://www.quirksmode.org/oddsandends/wbr.html

No dramas though, it was fine as-is and even better if it is working without.

File: src/Epic/index.tsx (hunk 4 above)
Comment (adamnfish) [reply src/Epic/index.tsx:95]: It feels a bit different to the React docs there because as well as being "children" it is also **before** other hard-coded children elements (like the `<span>` and the paragraphs. No harm if this makes React sense though, very happy to defer to react expertise!

File: src/EpicWithHeaderImage/index.tsx (hunk 1 above)
Comment (tjmw) [reply src/EpicWithHeaderImage/index.tsx]: Thanks for this! Ahh quirksmode, I haven't thought about that site for ages. I used to refer to it all the time.

File: src/Epic/index.tsx (hunk 4 above)
Comment (tjmw) [reply src/Epic/index.tsx]: I've update this, the prop in the `Epic` component is now called `headerSection`. So the new component implementation basically looks like this:

```typescript
export const EpicWithSpecialHeader: React.FC<EpicProps> = (props: EpicProps) => (
    <Epic {...props} headerSection={<HeaderSection />} />
);
```

File: src/BrazeEndOfArticleComponent.ts (hunk 2 above)
Comment (tjmw) [reply src/BrazeEndOfArticleComponent.ts]: I've gone with `EpicWithSpecialHeader` for the name, which I think is a slight improvement on `EpicWithHeaderImage`.
